resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsEventHandler.scala

```diff
@@ -0,0 +1,229 @@
+import org.apache.spark.deploy.k8s.Constants._
+import org.apache.spark.deploy.k8s.KubernetesConf
+import org.apache.spark.internal.Logging
+import org.apache.spark.scheduler.ExecutorExited
+import org.apache.spark.util.Utils
+
+private[spark] class ExecutorPodsEventHandler(
+    conf: SparkConf,
+    executorBuilder: KubernetesExecutorBuilder,
+    kubernetesClient: KubernetesClient,
+    eventProcessorExecutor: ScheduledExecutorService) extends Logging {
+
+  import ExecutorPodsEventHandler._
+
+  private val EXECUTOR_ID_COUNTER = new AtomicLong(0L)
+
+  private val totalExpectedExecutors = new AtomicInteger(0)
+
+  private val eventQueue = new LinkedBlockingQueue[Seq[Pod]]()
+
+  private val podAllocationSize = conf.get(KUBERNETES_ALLOCATION_BATCH_SIZE)
+
+  private val podAllocationDelay = conf.get(KUBERNETES_ALLOCATION_BATCH_DELAY)
+
+  private val kubernetesDriverPodName = conf
+    .get(KUBERNETES_DRIVER_POD_NAME)
+    .getOrElse(throw new SparkException("Must specify the driver pod name"))
+
+  private val driverPod = kubernetesClient.pods()
+    .withName(kubernetesDriverPodName)
+    .get()
+
+  // Use sets of ids instead of counters to be able to handle duplicate events.
+
+  // Executor IDs that have been requested from Kubernetes but are not running yet.
+  private val pendingExecutors = mutable.Set.empty[Long]
+
+  // We could use CoarseGrainedSchedulerBackend#totalRegisteredExecutors here for tallying the
+  // executors that are running. But, here we choose instead to maintain all state within this
+  // class from the persecptive of the k8s API. Therefore whether or not this scheduler loop
+  // believes an executor is running is dictated by the K8s API rather than Spark's RPC events.
+  // We may need to consider where these perspectives may differ and which perspective should
+  // take precedence.
+  private val runningExecutors = mutable.Set.empty[Long]
+
+  private var eventProcessorFuture: Future[_] = _
+
+  def start(applicationId: String, schedulerBackend: KubernetesClusterSchedulerBackend): Unit = {
+    require(eventProcessorFuture == null, "Cannot start event processing twice.")
+    logInfo(s"Starting Kubernetes executor pods event handler for application with" +
+      s" id $applicationId.")
+    val eventProcessor = new Runnable {
+      override def run(): Unit = {
+        Utils.tryLogNonFatalError {
+          processEvents(applicationId, schedulerBackend)
+        }
+      }
+    }
+    eventProcessorFuture = eventProcessorExecutor.scheduleWithFixedDelay(
+      eventProcessor, 0L, podAllocationDelay, TimeUnit.MILLISECONDS)
+  }
+
+  def stop(): Unit = {
+    if (eventProcessorFuture != null) {
+      eventProcessorFuture.cancel(true)
+      eventProcessorFuture = null
+    }
+  }
+
+  private def processEvents(
+      applicationId: String, schedulerBackend: KubernetesClusterSchedulerBackend) {
+    val currentEvents = new java.util.ArrayList[Seq[Pod]](eventQueue.size())
+    eventQueue.drainTo(currentEvents)
```

Comment:
Why do you need to drain the queue? Can you just dequeue and process events one by one sequentially?